[REACT-SDK] add multiprovider support (#494)
configcat/docs: website/docs/sdk-reference/react.mdx

Also, we recommend using [confidential targeting comparators](../targeting/targeting-rule/user-condition.mdx#confidential-text-comparators) in the Targeting Rules of those feature flags that are used in the frontend/mobile SDKs.

## Using multiple providers

> This section explains how can you use multiple providers in one application with ReactSDK.

Reviewer: ```suggestion
```

> It can be real life scenario when you arrange your feature flags into multiple configs. These configs have different sdkKeys and in your code you have to differentiate those with ConfigCatProvider's attribute called `id`.

Reviewer: To access a feature flag's value, your code needs to refer to that Config where that particular feature flag is located. This is what a `ConfigCatProvider` does for you; it refers to the Config by specifying the Config's `sdkKey`.

If your code needs access to feature flags from multiple Configs, you will need multiple `ConfigCatProvider`s - each pointing to a Config by specifying that particular Config's `sdkKey`.

And when your code wants to access a particular feature flag, it should refer to the specific `ConfigCatProvider` that is associated with the Config where that particular feature flag is located. This is exactly what the `providerId` allows your code to do: to refer to the specific `ConfigCatProvider` that is associated with the feature flag it wants to access.

> When you would like to use the proper instance you must specify the `id` value. The consumer side this argument's name is `providerId`.

Reviewer: ```suggestion
```

Also, we recommend using [confidential targeting comparators](../targeting/targeting-rule/user-condition.mdx#confidential-text-comparators) in the Targeting Rules of those feature flags that are used in the frontend/mobile SDKs.

## Using multiple providers

To access a feature flag's value, your code needs to refer to that Config where that particular feature flag is located. This is what a ConfigCatProvider does for you; it refers to the Config by specifying the Config's sdkKey.

If your code needs access to feature flags from multiple Configs, you will need multiple ConfigCatProviders - each pointing to a Config by specifying that particular Config's sdkKey.

> And when your code wants to access a particular feature flag, it should refer to the specific ConfigCatProvider that is associated with the Config where that particular feature flag is located. This is exactly what the providerId allows your code to do: to refer to the specific ConfigCatProvider that is associated with the feature flag it wants to access.

Reviewer: Nekem Ádámé jobban tetszik, mint a saját javaslatom.

Én magamtól a kötőszavakat írnám csak más sorrendbe, valahogy így.
Ha tetszik bármi az enyémből, akkor szabadon felhasználható.
De ha maarad Ádámé, nekem az is teljesen jó.
Szabin vagyok, légyszi ne várjatok az approve-omra.

`ConfigCatProvider` allows you to access the feature flags of a specific Config, identified by the `sdkKey` parameter. (More precisely, the `sdkKey` identifies a combination of a Config and its Environment.) However, in some cases, you may want to access feature flags from multiple Configs in your application. To do this, you will need to use multiple `ConfigCatProvider`s — each associated with a specific Config by providing the corresponding `sdkKey`.

Keep in mind that `useFeatureFlag` calls will default to using the nearest `ConfigCatProvider`, similar to how React's [useContext](https://react.dev/learn/passing-data-deeply-with-context) works. When working with multiple providers, it may be necessary to differentiate between them. You can achieve this by assigning a unique `id` to each provider (note that the absence of an `id` also acts as a unique identifier).

By passing the appropriate provider `id` in your `useFeatureFlag`, `useConfigCatClient`, or similar calls, you can explicitly specify which provider (i.e., which `sdkKey`, and ultimately which Config and Environment) to use.

> And when your code wants to access a particular feature flag, it should refer to the specific ConfigCatProvider that is associated with the Config where that particular feature flag is located. This is exactly what the providerId allows your code to do: to refer to the specific ConfigCatProvider that is associated with the feature flag it wants to access.

Reviewer 2: Köszi a tippeket! 👍 

A jó részét átvezettem a javaslatomba, néhány dolgot leszámítva:

> the `sdkKey`
Amikor kódbeli azonosítóra hivatkozunk, akkor két fajta stílust szoktunk használni: "the `X` parameter" vagy simán `X` névelő nélkül. A "the `X`" nekem nagyon furcsán hat. Ha nem a kódbeli azonosítóról beszélünk, akkor persze semmi gond, viszont akkor írjunk "the SDK Key"-t.

> a combination of a Config and its Environment
A Config és az Environment "merőleges" dolgok, egymástól függetlenül léteznek, az előbbi nem birtokolja az utóbbit. Persze értem, hogy miről van szó, de inkább ne ültessünk ilyen gondolatokat a user fejébe.

> By passing the appropriate
Itt meghagynám a Then-t, mert ez az időbeli sorrendjére utal az említett dolgoknak. Normál esetben először id-kat rendel a providerekhez, aztán hivatkozik ezekre a `useFeatureFlag` hívásokban.

```tsx

> const CC_PROVIDERID = {
>   COMMON: "COMMON",
>   FRONTEND: "FRONTEND"};

Reviewer 2: ```suggestion
const CC_PROVIDERID = {
  COMMON: "COMMON",
  FRONTEND: "FRONTEND"
} as const;
```

See also: https://maxheiber.medium.com/alternatives-to-typescript-enums-50e4c16600b1#ecf7

> <ConfigCatProvider sdkKey={"sdkKey-01"} id={CC_PROVIDERID.COMMON}> 
>   <ConfigCatProvider sdkKey={"sdkKey-02"} id={CC_PROVIDERID.FRONTEND}>

Reviewer 2: Let's use similar SDK Keys to what are used in other parts of the docs.

```suggestion
<ConfigCatProvider sdkKey="YOUR_SDK_KEY" id={CC_PROVIDERID.COMMON}> 
  <ConfigCatProvider sdkKey="YOUR_OTHER_SDK_KEY" id={CC_PROVIDERID.FRONTEND}>
```

      ...
      const myFeatureInCommon = useFeatureFlag("myFeatureInCommon", false, void 0, CC_PROVIDERID.COMMON)

      const myFeatureInFrontEnd = useFeatureFlag("myFeatureInFrontEnd", false, void 0, CC_PROVIDERID.FRONTEND)
      ...
  </ConfigCatProvider>
</ConfigCatProvider>
```

> When you are using multiple providers and they are not embed in each other it is not necessary to specify the `id` attribute:

Reviewer 2: ```suggestion
When you are using multiple providers and they are not nested, it is not necessary to specify the `id` attribute (as in this case it is clear to the `useFeatureFlag` calls which `ConfigCatProvider` they should use):
```

```tsx

> <ConfigCatProvider sdkKey={"sdkKey-01"}>
>   ...
>   const myFeatureInCommon = useFeatureFlag("myFeatureInCommon", false);
>   ...
> </ConfigCatProvider>
> 
> <ConfigCatProvider sdkKey={"sdkKey-02"}>

Reviewer 2: ```suggestion
<ConfigCatProvider sdkKey="YOUR_SDK_KEY">
  ...
  const myFeatureInCommon = useFeatureFlag("myFeatureInCommon", false);
  ...
</ConfigCatProvider>

<ConfigCatProvider sdkKey="YOUR_OTHER_SDK_KEY">
```

  ...
  const myFeatureInFrontEnd = useFeatureFlag("myFeatureInFrontEnd", false);
  ...
</ConfigCatProvider>
```

> If you are using higher order components you can set `providerId` parameter in `withConfigCatClient` function.

Reviewer 2: ```suggestion
If you are using higher order components, you can specify the `providerId` parameter in the `withConfigCatClient` function.
```

Reviewer 2: The missing comma after the if clause is a grammatical error. See https://www.ego4u.com/en/cram-up/writing/comma?10

Also, please avoid referencing to identifiers like "set `providerId` parameter". It should be either "set the `providerId` parameter" or simply "set `providerId`".

I'll update my suggestion accordingly.
